### PR TITLE
Fix mem leak with multiple listeners being registered

### PR DESCRIPTION
**Pull request by zupat**
re: #5 

If the widget is already listening to a ValueNotifier, don't register a new listener.

### CODE REVIEW

**Comment by lukehutch** [comment]
Thanks, good catch, but this still has a leak, because the set will keep growing over time if the listener is not called. Probably we need to use a `List<WeakReference>` instead, and iterate through the entire list each time `reactiveValue` is called, removing references that have become null, and checking if the context is in the list at the same time. That will give `reactiveValue` O(N) performance relative to the number of usages of it, but I think this is unavoidable. Any ideas?

**Comment by zupat** [comment]
I had something like this:

```dart
void _cleanUp() {
  final now = DateTime.now();

  // only clean up if there are more than 10 subscribers
  // and it's been more than 10 seconds
  if (_subscribers.length < 10 || now.difference(_lastPurge) < _k10seconds) {
    return;
  }
  _lastPurge = now;

//...do clean up

}
```

and ran clean up everytime a ValueNotifier is watched. Here is the full code in my package:

https://github.com/jinyus/dart_beacon/blob/cb8667b47e8f728edb13869385de01f789bd0d3d/state_beacon/lib/src/extensions/flutter.dart

**Comment by zupat** [comment]
I decided to use the approach in this PR because the worst case is that the listener stays alive if the value is never modified. I'm not sure it's a big enough deal to bear the runtime cost of cleaning up on every call...Maybe I should revert the change in my pkg.

**Comment by lukehutch** [comment]
OK, this is what I have -- I think it's the best that can be done. It still has O(N) performance in the number of listening elements per value, but I think it's the best that can be done. I had to change this from an extension to a subclass to avoid allocating a static map.

Let me know what you think:

```dart
import 'package:flutter/material.dart';
import 'package:flutter/scheduler.dart';

// Can't refer to listener while it is being declared, so we need this class
// to add a layer of indirection.
class _ListenerWrapper {
  void Function()? listener;
}

/// Extend [ValueNotifier] so that [Element] objects in the build tree can
/// respond to changes in the value.
class ReactiveValueNotifier<T> extends ValueNotifier<T> {
  ReactiveValueNotifier(super.value);

  final _subscribedElements = <int, WeakReference<Element>>{};
  var _subscribedElementIdx = 0;
  final Finalizer<void Function()> _finalizer = Finalizer((fn) => fn());

  /// Fetch the [value] of this [ValueNotifier], and subscribe the element
  /// that is currently being built (the [context]) to any changes in the
  /// value.
  T reactiveValue(BuildContext context) {
    final element = context as Element;
    if (_subscribedElements.values.map((e) => e.target).contains(element)) {
      // If element is already subscribed, return the current value without
      // subscribing again
      return value;
    }
    // Store a weak weak reference to the element in the map
    final elementRef = WeakReference(element);
    // Give each new subscription a unique index
    final elementIdx = _subscribedElementIdx++;
    _subscribedElements[elementIdx] = elementRef;
    // Tell the finalizer to remove the weak reference from the map
    // if the element is garbage collected
    _finalizer.attach(element, () => _subscribedElements.remove(elementIdx),
        detach: element);

    final listenerWrapper = _ListenerWrapper();
    listenerWrapper.listener = () {
      assert(
          SchedulerBinding.instance.schedulerPhase !=
              SchedulerPhase.persistentCallbacks,
          'Do not mutate state (by setting the value of the ValueNotifier '
          'that you are subscribed to) during a `build` method. If you need '
          'to schedule a value update after `build` has completed, use '
          '`SchedulerBinding.instance.scheduleTask(updateTask, Priority.idle)`, '
          '`SchedulerBinding.addPostFrameCallback(updateTask)`, '
          'or similar.');
      // If the element has not been garbage collected (causing
      // `elementRef.target` to be null), or unmounted
      if (elementRef.target?.mounted ?? false) {
        // Mark the element as needing to be rebuilt
        elementRef.target!.markNeedsBuild();
      }
      // Remove the listener -- only listen to one change per `build`
      removeListener(listenerWrapper.listener!);
      // Detatch the finalizer from the element, and remove it from the map
      _finalizer.detach(element);
      _subscribedElements.remove(elementIdx);
    };
    addListener(listenerWrapper.listener!);
    return value;
  }

  /// Use this method to notify listeners of deeper changes, e.g. when a value
  /// is added to or removed from a set which is stored in the value of a
  /// `ValueNotifier<Set<T>>`.
  void notifyChanged() {
    // ignore: invalid_use_of_protected_member, invalid_use_of_visible_for_testing_member
    notifyListeners();
  }
}
```

**Comment by zupat** [comment]
I like this approach...wouldn't it be better to calculate the map key by combining the element and notifier hashes? That way the lookup would be constant time. If not, you could use `iterable.any` instead of `map + contains`:

```dart
// _subscribedElements.values.map((e) => e.target).contains(element)
_subscribedElements.values.any((e) => e.target == element)
```

idea:
```dart
final key = Object.hashAll([element,this])
if (_subscribedElements.contains(key)) {
  // If element is already subscribed, return the current value without
  // subscribing again
  return value;
}
```

edit:

It also seems that dart doesn't guarantee that the Finalizer's callback is called. I've never used it so I'm not sure if this is a problem in practice.
> No promises are made that the callback will ever be called
https://api.flutter.dev/flutter/dart-core/Finalizer-class.html

**Comment by lukehutch** [comment]
No, using a hash value as a hash key is not a good idea, because of the chance of hash collisions. The chance may be small, but it would be bad practice to take that risk.

`Object.hashAll` is also not a good idea, because who knows what fields are going to be part of `Element` -- including but not limited to the fact that there may be some state fields that are hashed differently each time, because the same element is changing state. I haven't looked at the fields, but it's not even worth looking, because the implementation could change anytime.

Finalizers are not guaranteed to be called for 2 reasons: (1) it could take arbitrarily long for an object to be GC'd; (2) on shutdown, the VM is not obligated to collect all the garbage before being shut down (and almost certainly won't). This would be a big problem if you were managing an external (FFI) resource.

Actually they created a different `NativeFinalizer` because of the second reason: https://github.com/dart-lang/sdk/issues/52436

But for this purpose, the use of `Finalizer` is sufficient, because this is pure Dart code, and as long as the object is in fact GC'd, the finalizer will be called.

In other words, the finalizer may not be called, but it will be called if the code is still running, and the object is GC'd.

The only think I don't like about this approach is that we're still stuck with an O(N) search every time `build` is called. This is bad practice, but it may be unavoidable. Anyway Flutter runs O(N) searches all the time in its `.of` calls (`Provider.of` etc.), so maybe this is OK.

Also, the code is completely untested, I just read the docs (and didn't really understand them! I filed an issue in `dart-lang/language` about that). I don't even know how to properly test this, because I have no idea if a GC can be manually triggered.

I also have very limited time to properly fix this right now -- if you're willing to help test this change, I'd appreciate it!

**Comment by zupat** [comment]
> because who knows what fields are going to be part of `Element` -- including but not limited to the fact that there may be some state fields that are hashed differently each time, because the same element is changing state. I haven't looked at the fields, but it's not even worth looking, because the implementation could change anytime.

This would mean that `_subscribedElements.values.map((e) => e.target).contains(element)` suffers from the same issue. As it stands it currently uses the default `Object.hashCode` but you're right...that could change anytime.

**Comment by lukehutch** [comment]
That's not what `.map` does, it has nothing to do with `Map` :)

https://www.darttutorial.org/dart-tutorial/dart-map-method

It's a functional programming concept. It sucks that we use the same word for two unrelated things.

**Comment by zupat** [comment]
Haha, I know that. I meant the line is comparing the element passed in with the elements stored in the `_subscribedElements` Map, and you were saying hashing it would be problematic because its state can change.

**Comment by lukehutch** [comment]
No, the key is an `int`, and that is cleaned up by `Finalizer`. The value is wrapped in `WeakReference`, so `e.target` can be garbage collected (which triggers the finalizer).

The `.map` is needed because you can't keep any non-weak reference to the element (so the element can't be used as a key at all). Well, there are a couple of `Map` libraries that allow this, but adding those as a dep seemed too heavyweight.

I might try and duplicate the logic from one of those libraries though...

**Comment by lukehutch** [comment]
Fixed in 2.0.0, using `Expando` instead. Thanks!

**Comment by zupat** [comment]
Minor query: Wouldn't you still need the finalizer in case the widget gets disposed and the value is never modified?

**Comment by lukehutch** [comment]
I was going to add that back, but I figured if the value changes or the ValueNotifier is disposed, the listener will go away anyway. So this is not really a leak, it's more of a ValueNotifier lifecycle issue. But maybe a finalizer should be used to reduce memory pressure, you're right. I'll add it back.

**Comment by lukehutch** [comment]
@jinyus I added the `Finalizer` back in v2.0.1. Please test extensively!

**Comment by zupat** [comment]
The finalizer is very hard to test because its behavior isn't deterministic. When testing it in a demo app, it seems to work as intended but can take 1-5s to trigger. This is the widget test.

```dart
testWidgets('should remove listeners when widget is disposed',
      (WidgetTester tester) async {
    final testCounter = ValueNotifier(0);
    final show = ValueNotifier(true);

    expect(testCounter.hasListeners, false);

    await tester.pumpWidget(
      MaterialApp(
        home: CounterColmn(counter: testCounter, show: show),
      ),
    );

    // Should have listeners after widget is built
    expect(testCounter.hasListeners, true);

    show.value = false;

    await tester.pumpAndSettle();
    // await tester.pumpWidget(const MaterialApp(home: SizedBox()));

    // wait for finalizer to run
    await tester.runAsync(() async {
     // tried up to 15s
      await Future.delayed(Duration(seconds: 2));
    });

    // listeners should be removed
    expect(testCounter.hasListeners, false);
  });




class CounterColmn extends StatelessWidget {
  const CounterColmn({super.key, required this.counter, required this.show});

  final ValueNotifier<int> counter;
  final ValueNotifier<bool> show;

  @override
  Widget build(BuildContext context) {
    return Column(
      children: show.reactiveValue(context)
          ? [
              Counter(counter: counter),
              Counter(counter: counter),
              Counter(counter: counter),
              Counter(counter: counter),
              Counter(counter: counter),
            ]
          : [Container()],
    );
  }
}

class Counter extends StatelessWidget {
  const Counter({super.key, required this.counter});

  final ValueNotifier<int> counter;

  @override
  Widget build(BuildContext context) {
    return Text(
      counter.reactiveValue(context).toString(),
      style: Theme.of(context).textTheme.headlineMedium!,
    );
  }
}
```

**Comment by lukehutch** [comment]
Perfect, thanks for testing it! I didn't notice any anomalies in my own testing. In theory it should work, but this is the first time I am using `Finalizer`, and the documentation is terrible, so I wasn't even sure if I was using it right.

These days I just use `ValueNotifier` with `ValueListenableBuilder`, so I don't have extensive usage of this flutter_reactive_value library anymore in my own projects. Please let me know if you run into any issues using this. Thanks for the help getting this fixed!
